src/shogun/optimization/FirstOrderSAGCostFunctionInterface.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#ifndef FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
+#define FIRSTORDERSAGCOSTFUNCTIONINTERFACE_H
+
+#include <stan/math.hpp>
+#include <shogun/mathematics/eigen3.h>
+#include <shogun/lib/config.h>
+#include <shogun/optimization/FirstOrderSAGCostFunction.h>
+#include <functional>
+#include <vector>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+
+using Eigen::Dynamic;
+using Eigen::Matrix;
+using stan::math::var;
+using std::function;
+
+namespace shogun
+{
+	/** @brief The first order stochastic cost function base class for
+	 * implementing
+	 *  The SAG Cost function
+	 *
+	 * The class gives the implementation used in first order stochastic
+	 * minimizers
+	 *
+	 * The cost function must be Written as a finite sample-specific sum of
+	 * cost.
+	 * For example, least squares cost function,
+	 * \f[
+	 * f(w)=\frac{ \sum_i{ (y_i-w^T x_i)^2 } }{2}
+	 * \f]
+	 * where \f$(y_i,x_i)\f$ is the i-th sample,
+	 * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+	 */
+	class FirstOrderSAGCostFunctionInterface : public FirstOrderSAGCostFunction
+	{
+	public:
+		/** Consturctor that initializes X, y, trainable_parameters, and the
+		 * cost_function */
+		FirstOrderSAGCostFunctionInterface(
+		    SGMatrix<float64_t>* X, SGMatrix<float64_t>* y,
+		    Matrix<var, Dynamic, 1>* trainable_parameters,
+		    Matrix<function<var(int32_t)>, Dynamic, 1>* cost_for_ith_point,
+		    function<var(Matrix<var, Dynamic, 1>*)>* total_cost);
+
+		/** Default constructor, use setter helpers to set X, y,
+		*trainable_parameters, and
+		*		cost_function
+		*/
+		FirstOrderSAGCostFunctionInterface(){};
+
+		/** Setter for the training data X */
+		virtual void set_training_data(
+		    SGMatrix<float64_t>* X_new, SGMatrix<float64_t>* y_new);
+
+		/** Setter for the trainable parameters of the cost function */
+		virtual void
+		    set_trainable_parameters(Matrix<var, Dynamic, 1>* new_params);
+
+		/** Setter for the cost function definition using stan */
+		virtual void set_ith_cost_function(
+		    Matrix<function<var(int32_t)>, Dynamic, 1>* new_cost_f);
+
+		/** Setter for the overall cost function */
+		virtual void set_cost_function(
+		    function<var(Matrix<var, Dynamic, 1>*)>* total_cost);
+
+		virtual ~FirstOrderSAGCostFunctionInterface();
+
+		/** Initialize to generate a sample sequence
+		 *
+		 */
+		virtual void begin_sample();
+
+		/** Get next sample
+		 *
+		 * @return false if reach the end of the sample sequence
+		 * */
+		virtual bool next_sample();
+
+		/** Get the SAMPLE gradient value wrt target variables
+		 *
+		 * WARNING
```

Comment:
not sure what the warning is supposed to mean, I commented this before but maybe I missed your reply?

Reply:
@karlnapf It means we should return the partial derivative of the ith sample with respect to the variables, and not the partial derivative of the total cost with respect with respect to the variables. 

So for example, if f_i(w) = (w^t*X[i]-y[i])^2 is the error of the ith data point, then this function returns the partial derivative of the **current** data sample with respect to w, rather than returning the derivative of Sum_i(f_i(w)) with respect to w.

This function (and get_average_gradient()) is from the base class (FirstOrderSAGCostFunction) and had the warning in its docs. Do you want me to change the warning in both files to something else, and if so, what do you want to clarify?